docs/rfc/rfc-1.adoc

```diff
@@ -253,12 +253,26 @@ wallet and closes.
 
 === Wallet Lifecycle
 
-Wallets are periodically created, where the period length is a governable
-parameter. To create a new wallet, a group of 100 operators is selected from
-the pool of available operators (some operators may be selected twice if there
-are not enough) using a process called sortition. The probabiliy that a
-particular operator is chosen is based on their stake weight, which in turn is
-based on the number of `T` tokens they have invested in the staking contract.
+Governable Parameters:
+
+- `wallet_dust_threshold` the smallest amount of btc a wallet can hold before we
+  attempt to close the wallet and transfer the funds to a randomly selected
+  wallet
+- `wallet_creation_period` how frequently we attempt to create new wallets
+- `wallet_max_btc` the amount of btc a wallet can hold before we trigger the
+  early creation of a new wallet
+- `max_wallet_age` the oldest we allow a wallet to become before we transfer the funds
+  to a randomly selected wallet
+
+A new wallet is created when either enough time has passed as defined in
+`wallet_creation_period` *AND* the wallet contains at least
+`wallet_dust_threshold` btc *OR* the wallet currently contains more btc than
+`wallet_max_btc`, whichever happens first. To create a new wallet, a group of
+100 operators is selected from the pool of available operators (some operators
+may be selected twice if there are not enough) using a process called
+sortition. The probabiliy that a particular operator is chosen is based on
+their stake weight, which in turn is based on the number of `T` tokens they
+have invested in the staking contract.
```

Comment:
What is triggering wallet creation? New deposit? Or is it some public function that will be called by maintainers (lucky or unlucky)?

Reply:
A new wallet needs to be created either when enough time has passed since the last one was created or when there's too much bitcoin in the current wallet, whichever comes first. We also need to kick that process off on-chain somehow (and will need to call the sortition pool), so that we can penalize folks if it doesn't start happening.

I think this is a great candidate for the mechanism where we offer an increasing reward for submitting these transactions until someone's client wants to do it since I don't think we want to slow wallet creation on the whole network down waiting for one operator to get their act together. Same deal for paying the gas to have the sortition pool pick the 100 folks for the wallet (since I imagine that'll be expensive).

Reply:
For sortition pool selection, we can use the same pattern as for the random beacon: 

> New groups should be created with a fixed, governable frequency of relay requests. Instead of a ticket-based approach for group selection, we should use sortition pools. Group creation start transaction should be embedded into relay request transaction and should lock a sortition pool. From this moment, no operator can enter or leave the pool. Once a new relay entry appears on the chain, all off-chain clients should perform group selection optimistically, using the new entry value and a view sortition pool function call. After determining group members, clients should perform DKG as usual. Once the group submits the result to the chain, a challenge period starts.
>
> During the challenge period, anyone can notify that the submitted DKG result contains group members not selected by the pool. This leads to slashing all members who signed the result and the notifier receives 5% from the slashed amount. The malicious result is immediatelly discarded. The length of the challenge period and slashing amount are governable parameters.
>
> Once the challenge period passes, anyone can unlock the sortition pool and mark the DKG result as accepted. In this transaction, the DKG result submitter receives a reward, as described in Fees and Rewards section.
>
> There is a timeout before which DKG result should be submitted. The timeout equals the group size multiplied by the number of blocks for a member to become eligible to submit DKG result. The timer starts at the moment when the first member becomes eligible.
>
> In case the DKG result was not submitted before a timeout, anyone can notify DKG timed out and receive a reward, as described in Fees and Rewards section. DKG timeout includes the situation when no new relay entry was produced and sortition could not be performed.

https://github.com/keep-network/keep-core/blob/main/docs/rfc/rfc-19-random-beacon-v2.adoc#group-creation

So in this case, that would be:
1. Someone notifies the contract to create a new wallet.
2. Sortition pool is locked, relay request is published.
3. Beacon produces relay entry, publishes it to the chain.
4. Clients perform a selection from the sortition pool off-chain calling sortition function.
5. Clients execute tECDSA key generation and publish the result.
6. Challenge period starts, anyone can challenge the result, especially group members in the published result.
7. If no challenge was published, anyone can call a transaction to approve the result. Usually, it's the original result submitter - they are incentivized to do so because in this transaction we release their reward. If they do not do that until some timeout, anyone can call the approve transaction and receive the submitter's reward.

In this model, we need three types of rewards:
- reward for one notifying a new wallet should be created
- reward for DKG result submitter
- reward for notifier of DKG timeout (pool needs to be unlocked)

For the first reward we can use the mechanism you described, for two other rewards, we can use just a plain token transfer of some governable amounts from some pool.

What I like about it is that we can reuse some code from the random beacon.

What is problematic though - and I think it deserves a separate PR - is how these rewards are funded. Who, when, and why. The easiest option would be - DAO from inflation, and maybe that's enough for v1 🤷🏻